### PR TITLE
feat: add packageImportMethod option

### PR DESCRIPTION
**Pull request by etamponi**
I didn't want to spend too much time figuring out the right signature for the `execFilePromise` function, but everything else should be there. As for the tests, in this case, I would have some troubles as reflinks only work on some kind of filesystems.

I can mock `child_process` if you want, using `mockery` or similar tools.

### CODE REVIEW

File: src/link/index.ts [hunk 1]
```diff
@@ -21,10 +21,31 @@ import linkIndexedDir from '../fs/linkIndexedDir'
 import ncpCB = require('ncp')
 import thenify = require('thenify')
 import {rootLogger, statsLogger} from '../loggers'
+import child_process = require('child_process')
+import util = require('util')
+let execFilePromise: any  // tslint:disable-line
+if (util.promisify) {
```
Comment (zkochan) [inline src/link/index.ts]: you can use `thenify`. It is already in the dependencies

**Review by zkochan — CHANGES_REQUESTED** [review]
Copying is already done when hardlinking fails (https://github.com/pnpm/supi/blob/master/src/link/index.ts#L252-L255).

There are `linkPkg()` and `copyPkg()`, they are passed in to `linkAllPkgs`. You should create a third funcion called probably `reflinkPkg`.

I think the best way to do it is to create an option called `packageImportMethod` (maybe you come up with a better name) with 4 available options:

* `auto` - the default one. Tries to hardlink packages. If it fails to do so, falls back to copying.
* `hardlink` - tries to hardlink, if it fails to do so, just fails.
* `copy` - copies files
* `reflink` - copies with `cp -r --reflink`

Is there a feature detection for `reflink`? You could write a test that would be skipped on systems that don't support it.

**Comment by zkochan** [comment]
Related: https://github.com/pnpm/pnpm/issues/771

**Comment by etamponi** [comment]
I tried to make as few changes as possible, but if the switch statement is too "clever", let me know, I'll change it.

**Comment by zkochan** [comment]
A switch is a good choice in this case but I would suggest declaring a separate function for creating the reflinks - `reflinkPkg`. And you can pass `reflinkPkg` to `linkAllPkgs()`. Same way as linkPkg and copyPkg are passed there. You can rename linkPkg to hardlinkPkg

File: src/link/index.ts [hunk 2]
```diff
@@ -246,13 +250,29 @@ async function linkNewPackages (
...
+          break
+        case 'auto':
+          try {
+            await linkAllPkgs(hardlinkPkg, newPkgs, opts)
+            break
+          } catch (err) {
+            if (!err.message.startsWith('EXDEV: cross-device link not permitted')) throw err
+            logger.warn(err.message)
+            logger.info('Falling back to copying packages from store')
+          }
```
Comment (zkochan) [inline src/link/index.ts]: I did not configure linting yet but it wouldn't allow a case w/o break, so better just duplicate the copy code here and use a break

**Comment by zkochan** [comment]
If you don't know how to cover reflink with a test, skip it for now, but let's create an issue for it.

For copying you can create a test, I believe

And please document the new option in the README

**Comment by etamponi** [comment]
I'll write the tests either later today or next year ;)

**Comment by etamponi** [comment]
Fun: the copy test is failing. Has the copy path been tested before? This is the error:

```
$ node_modules/.bin/ts-node test/packageImportMethods.ts                                                                                                                   ⬡ 6.12.2 [±master-upstream ●●]
TAP version 13
# packageImportMethod can be set to copy
ok 1 create testing package 1
not ok 2 Error: ENOENT: no such file or directory, mkdir '/Users/etamponi/code/.tmp/1/project/node_modules/.localhost+4873/is-negative/2.1.0/node_modules/is-negative'
  ---
    operator: error
    expected: |-
      undefined
    actual: |-
      [ { [Error: ENOENT: no such file or directory, mkdir '/Users/etamponi/code/.tmp/1/project/node_modules/.localhost+4873/is-negative/2.1.0/node_modules/is-negative'] errno: -2, code: 'ENOENT', syscall: 'mkdir', path: '/Users/etamponi/code/.tmp/1/project/node_modules/.localhost+4873/is-negative/2.1.0/node_modules/is-negative' } ]
    at: bound (/Users/etamponi/code/supi/node_modules/.registry.npmjs.org/onetime/2.0.1/node_modules/onetime/index.js:30:12)
    stack: |-
      Error: Error: ENOENT: no such file or directory, mkdir '/Users/etamponi/code/.tmp/1/project/node_modules/.localhost+4873/is-negative/2.1.0/node_modules/is-negative'
          at Test.assert [as _assert] (/Users/etamponi/code/supi/node_modules/.registry.npmjs.org/tape/4.8.0/node_modules/tape/lib/test.js:212:54)
          at Test.bound [as _assert] (/Users/etamponi/code/supi/node_modules/.registry.npmjs.org/tape/4.8.0/node_modules/tape/lib/test.js:64:32)
          at Test.error.Test.ifError.Test.ifErr.Test.iferror (/Users/etamponi/code/supi/node_modules/.registry.npmjs.org/tape/4.8.0/node_modules/tape/lib/test.js:332:10)
          at Test.bound [as ifError] (/Users/etamponi/code/supi/node_modules/.registry.npmjs.org/tape/4.8.0/node_modules/tape/lib/test.js:64:32)
          at Test.end (/Users/etamponi/code/supi/node_modules/.registry.npmjs.org/tape/4.8.0/node_modules/tape/lib/test.js:137:14)
          at bound (/Users/etamponi/code/supi/node_modules/.registry.npmjs.org/tape/4.8.0/node_modules/tape/lib/test.js:64:32)
          at bound (/Users/etamponi/code/supi/node_modules/.registry.npmjs.org/onetime/2.0.1/node_modules/onetime/index.js:30:12)
  ...
```

**Comment by zkochan** [comment]
Maybe it is this issue? https://github.com/pnpm/pnpm/issues/866

**Comment by etamponi** [comment]
I don't know, but an `await mkdirp(dependency.hardlinkedLocation)` before the call to `ncp` fixed it :)

**Comment by zkochan** [comment]
wow, if it will fix that issue, you are the hero of the day!

**Comment by etamponi** [comment]
Here you go! I've added the copy test. I'd love to at least unit-test the reflink method using a mock of `child_process`, but if you're ok with it, we can do it later. I'll open a new issue

**Comment by zkochan** [comment]
I am ok with later. I would prefer the real thing. Mocks are not that reliable

**Comment by etamponi** [comment]
The problem with the real thing is that reflinks only work on btrfs and ocfs2...

File: src/link/index.ts [hunk 3]
```diff
@@ -360,12 +385,30 @@ async function linkAllModules (
...
   opts: {
     force: boolean,
     baseNodeModules: string,
+    packageImportMethod: 'auto' | 'hardlink' | 'copy' | 'reflink',
+  }
+) {
+  const pkgJsonPath = path.join(dependency.hardlinkedLocation, 'package.json')
+
+  if (!filesResponse.fromStore || opts.force || !await exists(pkgJsonPath)) {
+    await execFilePromise('mkdir', ['-p', dependency.hardlinkedLocation])
```
Comment (zkochan) [inline src/link/index.ts]: You can use the mkdirp lib here. It is more reliable because it retries on fails.

**Review by zkochan — APPROVED** [review]
LGTM, added just one comment

**Pull request merged**